src/gluonts/mx/distribution/isqf.py

```diff
@@ -0,0 +1,966 @@
+        )
+
+        spline_alpha_tilde = self.cdf_spline(z)
+
+        for i in range(self.num_qk - 1):
+
+            is_in_between = F.broadcast_logical_and(
+                F.slice_axis(qk_y, axis=-1, begin=i, end=i + 1).squeeze(-1)
+                <= z,
+                z
+                < F.slice_axis(qk_y_plus, axis=-1, begin=i, end=i + 1).squeeze(
+                    -1
+                ),
+            )
+
+            alpha_tilde = F.where(
+                is_in_between,
+                F.slice_axis(
+                    spline_alpha_tilde, axis=-1, begin=i, end=i + 1
+                ).squeeze(-1),
+                alpha_tilde,
+            )
+
+        return alpha_tilde
+
+    def sample(
+        self, num_samples: Optional[int] = None, dtype=np.float32
+    ) -> Tensor:
+        r"""
+        Function used to draw random samples
+        Parameters
+        ----------
+        num_samples
+            number of samples
+        dtype
+            data type
+        Returns
+        -------
+        Tensor
+            Tensor of shape (*batch_shape,) if num_samples = None else (num_samples, *batch_shape)
+        """
+
+        F = self.F
+
+        # if num_samples=None then input_alpha should have the same shape as beta_l, i.e., (*batch_shape,)
+        # else u should be (num_samples, *batch_shape)
+        alpha = F.random.uniform_like(
+            data=(
+                self.beta_l
+                if num_samples is None
+                else self.beta_l.expand_dims(axis=0).repeat(
+                    axis=0, repeats=num_samples
+                )
+            )
+        )
+
+        sample = self.quantile(alpha)
+
+        if num_samples is None:
+            sample = F.squeeze(sample, axis=0)
+
+        return sample
+
+    @property
+    def batch_shape(self) -> Tuple:
+        return self.beta_l.shape
+
+    @property
+    def event_shape(self) -> Tuple:
+        return ()
+
+    @property
+    def event_dim(self) -> int:
+        return 0
+
+
+class ISQFOutput(DistributionOutput):
+    r"""
+    DistributionOutput class for the Incremental (Spline) Quantile Function
+
+    Parameters
+    ----------
+    num_pieces
+        number of spline pieces for each spline
+        ISQF reduces to IQF when num_pieces = 1
+    alpha
+        Tensor containing the x-positions of quantile knots
+    tol
+        tolerance for numerical safeguarding
+    """
+
+    distr_cls: type = ISQF
+
+    @validated()
+    def __init__(
+        self, num_pieces: int, qk_x: List[float], tol: float = 1e-4
+    ) -> None:
+        # ISQF reduces to IQF when num_pieces = 1
+
+        super().__init__(self)
+
+        assert (
+            isinstance(num_pieces, int) and num_pieces > 0
+        ), "num_pieces should be an integer and greater than 0"
+
+        self.num_pieces = num_pieces
+        self.qk_x = sorted(qk_x)
+        self.num_qk = len(qk_x)
+        self.tol = tol
+        self.args_dim = cast(
+            Dict[str, int],
```

Comment:
Not sure cast is necessary here. You could probably also just do `self.args_dim: Dict[str, int] = ...`